### ISSUE TITLE
Bulk Update for MySQL provider

### ISSUE THREAD

**Issue by comik** [linq2db/linq2db#1259]
Any chance to have bulk update for MySQL provider in nearest future?
Highly desired function.

Thanks!

**Comment by sdanyliv** [comment]
It should work for MySql.
At first you have to BulkCopy records into temporary table and then create update query. See my comments here 
https://github.com/linq2db/linq2db/issues/1252#issuecomment-408617968
If it is not working I’ll catch this issue.

**Comment by comik** [comment]
It probably works if you know the type you're working with and can map values accordingly. But I'm trying to build a repository class with generic UpdateRange method.
Any thoughts?

**Comment by sdanyliv** [comment]
Oh no, generic repository again ;)
In this case you have to invoke and create linq expression chain by yourself. It is not easy but possible.

**Comment by sdanyliv** [comment]
I’m not a fan of repository pattern, even worse i think it is worst approach after inventing linq and proper linq provider were written )
My thoughts about this are here:
https://github.com/linq2db/linq2db/issues/848#issuecomment-328525236
But looks like you stuck with this approach and maybe i’ll find time to write dynamic sample.
So you have a list of objects that needs to be updated in bulk by all fileds that are not keys or non-updatable?

**Comment by sdanyliv** [comment]
How many records do you update usually?

**Comment by comik** [comment]
Between 1 and 3k. But max count can be limited.

**Comment by sdanyliv** [comment]
I can’t believe that you update all fields for all 3k records. looks like you just updating one, maximum 2-3 fileds, like `IsDeleted = true` or `SomeCounter = SomeCounter + 1`.
If it is true maybe we can find much better approach?

**Comment by comik** [comment]
Yes, it is true. Generally, we're updating 1 to 5 fields per row.

**Comment by sdanyliv** [comment]
It is unit tested by some mocks? Can you avoid this? Just trust our code?
Also does these updates use just previous value of the same record or some parameter? 
Any additional antipatterns like UnitOfWork?

**Comment by comik** [comment]
Tested, but can be avoided. Not sure what you mean by "use just previous value of the same record". We read data from DB, update some fields and then update those rows (matched by PK). No other (anti)patterns.
Database access code usually looks like:
`using (var repoFactory = new RepoFactory())
{
    var repo = repoFactory.CreateFor<Entity>();
    var list = repo.FindMatching(e => e.Type == 1);
    // modifying data here
    repo.UpdateRange(list);
}`

**Comment by sdanyliv** [comment]
> Not sure what you mean by "use just previous value of the same record"

If you need just increment counter or set values for all records it is better to use our extensions.

Ok, i’ve got the point. How about specifying which fileds have to be updated?
Or all updatable is ok?

**Comment by comik** [comment]
All updatable should be ok.

**Comment by comik** [comment]
Hey, Svyatoslav!
I'm not sure if you're connected to BLToolkit in any way (as I consider it to be ancestor of linq2db, right?), but you maybe you have any suggestions on issue https://github.com/igor-tkachev/bltoolkit/issues/405?
Thanks in advance!

**Comment by sdanyliv** [comment]
@comik, i have prepared function for you. Currently it is in dev branch but someday we will publish it in other library because it needs production ready polishing.

https://github.com/linq2db/linq2db/blob/dev-batch_update/Tests/Linq/UserTests/DynamicExtensions.cs#L197

And sample usage

https://github.com/linq2db/linq2db/blob/dev-batch_update/Tests/Linq/UserTests/BulkUpdateTests.cs#L94

**Comment by darkflame0** [comment]
> @comik, i have prepared `BulkUpdate` function for you. Currently it is in dev branch but someday we will publish it in other library because it needs production ready polishing.
> 
> https://github.com/linq2db/linq2db/blob/dev-batch_update/Tests/Linq/UserTests/DynamicExtensions.cs#L197
> 
> And sample usage
> 
> https://github.com/linq2db/linq2db/blob/dev-batch_update/Tests/Linq/UserTests/BulkUpdateTests.cs#L94

any progress? I cant find it In the current version

**Comment by viceroypenguin** [comment]
There is not yet a BulkUpdate() method specifically, but TempTables have been added, as well as Merge statements. I would do something similar to this:

```csharp
await using var tmp = await context.CreateTempTableAsync<Table>();
await tmp.BulkCopyAsync(data);
await destination.Merge().Using(tmp)
  .On((dst, src) => dst.Key == src.Key)
  .InsertWhenNotMatched(src => new Destination { Key = src.Key, ColA = src.ValueA, })
  .UpdateWhenMatched((dst, src) => new Destination { ColA = src.ValueA, })
  .MergeAsync();
```

**Comment by darkflame0** [comment]
I think it's still a little complicated...
In most cases, we just need batch update instead of merge

**Comment by MaceWindu** [comment]
@viceroypenguin MySQL doesn't have merge

**Comment by ilsnk** [comment]
is it relevant? when can you release it in production version?
